### PR TITLE
Use a global interrupt handler and handle interrupt edge cases

### PR DESCRIPTION
**Pull request by seeM**
Attempts to address #3805.

There are two parts to this:

1. Moved from using cell-level interrupt (`NotebookCellExecution.token`) to notebook-level interrupt (`NotebookController.interruptHandler`), which I believe is the correct approach when there is really a single global kernel backing executions. Cell-level interrupt also only let you interrupt once, and sometimes Python requires multiple interrupts.
2. Better handling for edge cases when interrupting such as the session having exited unexpectedly.

### QA Notes

Since the addressed issue is intermittent, it's hard to test. Opening and closing notebooks, and starting and interrupting executions in different ways should all be more stable with this PR.

Integration tests should also pass locally.

### CODE REVIEW

**Review by nstrayer — APPROVED** [review]
Looks good. Since it's hard to reproduce exactly I just verified that the code seems correct and the correct paths are being taken when running.

File: extensions/positron-notebook-controllers/src/notebookController.ts [hunk 1]
```diff
@@ -148,11 +189,20 @@ export class NotebookController implements vscode.Disposable {
...
 		// Queue all cells for execution; catch and log any execution errors.
 		try {
-			await Promise.all(cells.map(cell => this.queueCellExecution(cell, notebook)));
+			await Promise.all(cells.map(cell => this.queueCellExecution(cell, notebook, tokenSource.token)));
 		} catch (err) {
 			log.debug(`Error executing cells: ${err}`);
+		} finally {
+			// Clean up the cancellation token source for this execution.
+			if (this._executionTokenSourceByNotebookUri.get(notebook.uri) === tokenSource) {
+				this._executionTokenSourceByNotebookUri.delete(notebook.uri);
```
[the remark this answers is not in the document]
Comment (seeM) [reply extensions/positron-notebook-controllers/src/notebookController.ts:204]: Exactly
